### PR TITLE
Dynamic endpoint version

### LINKED ISSUE #4120: Allow for dynamic selection of WP.com endpoint version
The version for WP.com endpoints is currently globally set and hard-coded.  The problem with this approach is that it's difficult to work on upcoming changes, until all services have been upgraded.  The intention of this change is not to start releasing code for beta versions of the API, but rather to allow a certain flexibility on upgrading our APIs.

Ideally a global endpoint version should be set, which can be overridden by services that need to.

### PR DESCRIPTION
**Pull request by diegoreymendez**
Resolves #4120 

The REST services now allow the caller to pick the API version to use.

This PR does not intend to promote the inclusion of beta versions of the API in the release app, but merely as a mechanism to allow developers to:

- Start developing and testing for incoming versions of the API (without releasing the new functionality until the API is released).
- Allow for partial and progressive migration to newly released API versions (as opposed to having to migrate all services simultaneously).

**Test 1:**
Run the unit tests and make sure none of them fail.

**Test 2:**
Navigate and use the app.  Try to make sure no service calls are failing.  (This would be easier to achieve if we had unit tests for all services, so that'll be an upcoming goal)

/cc @astralbodies, @aerych, @jleandroperez, @koke, @SergioEstevao, @sendhil 

(I want as many eyes as possible here... been testing extensively and fixed all the issues I could find)

### CODE REVIEW

File: WordPress/Classes/Networking/ServiceRemoteREST.h [hunk 1]
```diff
@@ -1,5 +1,12 @@
 #import <Foundation/Foundation.h>
 
+typedef enum
+{
+    ServiceRemoteRESTApiVersion_1_1 = 1001,
+    ServiceRemoteRESTApiVersion_1_2 = 1002,
+    ServiceRemoteRESTApiVersionDefault = ServiceRemoteRESTApiVersion_1_1,
+} ServiceRemoteRESTApiVersion;
```
Comment (koke) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: Why not use the constant strings directly instead of having to `switch` an enum to get them?

**Comment by koke** [comment]
There's a lot of new `requestUrlForDefaultApiVersionAndResourceUrl`, I feel like code using the default API version shouldn't need to change.

Some ideas:
- Add new `GET/POST/PUT/DELETE` methods that take a version parameter
```objc
[self.api version:ServiceRemoteRESTApiVersion_1_2
              GET:path
       parameters:parameters
          success:success
          failure:failure];
```

- Leave `WordPressComApiClientEndpointURL` with the default version number, and use absolute paths on requests that need a specific version

```objc
NSString *path = [self requestUrlForApiVersion:ServiceRemoteRESTApiVersion_1_2
                                   resourceUrl:@"read/menu"]; // => @"/rest/v1.2/read/menu"
[self.api GET:path parameters:parameters success:success failure:failure];
```

* I had another one that would look nice IMO, but seems way too complicated to implement:

```objc
[[self.api version:version:ServiceRemoteRESTApiVersion_1_2]
 GET:path
 parameters:parameters
 success:success
 failure:failure];
```

**Comment by diegoreymendez** [comment]
@koke - I went through the same thoughts, and agree to a certain extent.  For instance, the repeated calls to `requestUrlForDefaultApiVersionAndResourceUrl:` kind of suck.  But...

...my final opinion is a bit different, though.  I believe there should be no default API version.  Each call should just add "v1.1", "v1.2", etc explicitly.

The reason is that each API call has a certain path and parameters composition that is guaranteed to work for a certain API version (starting or ending at that version).  In other words, the call composition is the one that defines what API version must be used - going for "default" would be an invitation to break stuff when the default API version is changed.  Upgrading to a new API version would mean we'd be able to only upgrade specific calls, and do the whole migration progressively.

On the other hand having a default means changing that default forces you to automatically check all the services that use such default.  This means lots of testing, lots of changes and big PRs, long reviews.

So my idea for the next steps is, even though the change is big: let's try to move our code to the place where we want it to be.  Thoughts?

**For example:**
What about replacing all calls to `requestUrlForDefaultApiVersionAndResourceUrl:` with `requestUrlForApiVersion:resourceUrl:` so each call has it's specific version explicitly set?

File: WordPress/Classes/Networking/ServiceRemoteREST.h (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: I originally had constant strings, but it felt weird to allow anything in (ie: the user could then pass `@"hello world"` as the version number).

**Comment by koke** [comment]
> I believe there should be no default API version

I think I agree with that. It sucks having to replace all that code, but I think it's safer overall.

> What about replacing all calls to requestUrlForDefaultApiVersionAndResourceUrl: with requestUrlForApiVersion:resourceUrl: so each call has it's specific version explicitly set?

I'd really like a syntax where you could add the version to the `GET` method directly. Maybe it's just the `requestUrl` bit throwing me off though. How about what you proposed but named `pathForEndpoint:withVersion:`?

**Comment by diegoreymendez** [comment]
@koke - Agreed, I'll make that change and ping you back when it's done.

File: WordPress/Classes/Networking/ServiceRemoteREST.h (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: Thoughts?

Comment (koke) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: I'd have no problem with that. Not that it would preven using an arbitrary string, but a typedef could at least hint the developer what's expected.

https://gist.github.com/koke/8f3c1fd3ab9faa53b40d

And you can always add a `NSParameterAssert` if an unexpected version is passed.

**Comment by diegoreymendez** [comment]
Tagging as "Not ready for review" again until I can apply the changes mentioned above.

File: WordPress/Classes/Networking/ServiceRemoteREST.h (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: I was just reading through the code and I think I'd still prefer to limit the amount of options available to an enum rather than accept strings as an input.  Accepting strings eliminates the need to map the values, but other than that it makes the parameter a bit more flexible than I'd like it to be.

It's really not that big of a difference but it has the added value of being much stricter in allowing any value outside the intended ones.

Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: PS: this was marked as outdated by some unrelated changes.

**Comment by diegoreymendez** [comment]
This is ready again for review.

File: WordPress/Classes/Networking/ServiceRemoteREST.m [hunk 2]
```diff
@@ -1,6 +1,10 @@
 #import "ServiceRemoteREST.h"
 #import "WordPressComApi.h"
 
+static NSString* const ServiceRemoteRESTApiVersionStringInvalid = @"invalid_api_version";
+static NSString* const ServiceRemoteRESTApiVersionString_1_1 = @"v1.1";
+static NSString* const ServiceRemoteRESTApiVersionString_1_2 = @"v1.2";
```
Comment (koke) [inline WordPress/Classes/Networking/ServiceRemoteREST.m]: This is OK, but just FYI, when you declare a `static` constant like this, it's only defined for the current file so there's no need to use a prefix to namespace it. This one could be just `ApiVersionString_1_2`

File: WordPress/WordPressApi/WordPressComApi.m [hunk 3]
```diff
@@ -7,7 +7,7 @@
 #import "NotificationsManager.h"
 #import "WPUserAgent.h"
 
-static NSString *const WordPressComApiClientEndpointURL = @"https://public-api.wordpress.com/rest/v1.1/";
+static NSString *const WordPressComApiClientEndpointURL = @"https://public-api.wordpress.com/rest/";
```
Comment (koke) [inline WordPress/WordPressApi/WordPressComApi.m]: I'm wondering what happens if one of the developers doesn't notice we've changed how we use `WordPressComApi` and calls, for instance, `GET:@"me"`. Something that used to work will return a 404. Perhaps we should add some assertions to the GET/POST/PUT/DELETE methods to check that the URL starts with `v%d`

**Comment by aerych** [comment]
`ServiceRemoteRESTApiVersion` is maybe not specific enough?  I'm thinking about the upcoming WP-API vs the wpcom REST API.  Just a thought.

File: WordPress/Classes/Networking/ServiceRemoteREST.m (hunk 2 above)
Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.m]: Even though it's static (thus locally scoped), we can still have name-clashes with other globally-scoped constants.

See this example declaration of a `static const`:
screen shot 2015-08-18 at 3 05 33 pm

The only difference with a globally-scoped constant is that `static` constants won't clash with other `static` constants.

File: WordPress/WordPressApi/WordPressComApi.m (hunk 3 above)
Comment (diegoreymendez) [inline WordPress/WordPressApi/WordPressComApi.m]: Well, I can definitely add those assertions, since this would be pretty fresh, and they'd help us debug this feature as well.  Good recommendation.

The dev should notice the problems when testing the service, though (since it'd fail).  But that's just a side-note... the assertions would be a nice addition.

**Comment by diegoreymendez** [comment]
@aerych - That's a good observation.  Certainly one to discuss.

However, for the scope of this PR, and considering we currently have a lot of code refering to the REST API as the WordPress.com REST API, I'd consider leaving that decision out for now.

The thing is, as soon as we start introducing the new REST Api we'll need to review our REST naming conventions across our whole codebase.

File: WordPress/WordPressApi/WordPressComApi.m (hunk 3 above)
Comment (diegoreymendez) [inline WordPress/WordPressApi/WordPressComApi.m]: This is now addressed - I added methods that (a bit primitively) verify that the version is present in the path.

**Comment by diegoreymendez** [comment]
@aerych, @koke - Let me know your thoughts on my comments.

**Comment by aerych** [comment]
> However, for the scope of this PR, and considering we currently have a lot of code refering to the REST API as the WordPress.com REST API, I'd consider leaving that decision out for now.

> The thing is, as soon as we start introducing the new REST Api we'll need to review our REST naming conventions across our whole codebase.

No worries!  It was just a thought. 

I don't have any other comments on the code in the patch. I've taken the branch for a test drive, paying particular attention to the reader, and didn't notice anything breaking. :shipit: from me :)

**Comment by diegoreymendez** [comment]
Thank you @aerych!

I'll wait for @koke's approval before merging this, so anything else, let me know.

File: WordPress/Classes/Networking/ServiceRemoteREST.m [hunk 4]
```diff
@@ -16,4 +20,40 @@ - (id)initWithApi:(WordPressComApi *)api {
...
+        case ServiceRemoteRESTApiVersion_1_1:
+            result = ServiceRemoteRESTApiVersionString_1_1;
+            break;
+            
+        case ServiceRemoteRESTApiVersion_1_2:
+            result = ServiceRemoteRESTApiVersionString_1_2;
+            break;
+            
+        default:
+            result = ServiceRemoteRESTApiVersionStringInvalid;
```
Comment (jleandroperez) [inline WordPress/Classes/Networking/ServiceRemoteREST.m]: Maybe an assertion here?

**Comment by jleandroperez** [comment]
@diegoreymendez i've just noticed `PushAuthenticationServiceRemote` lacks the new dynamic versioning. Was that one left out intentionally?

**Comment by diegoreymendez** [comment]
@jleandroperez - Only that I didn't notice it.  I'm adapting it now... what's a good test case to trigger that service?

File: WordPress/Classes/Networking/ServiceRemoteREST.m (hunk 4 above)
Comment (diegoreymendez) [inline WordPress/Classes/Networking/ServiceRemoteREST.m]: :+1:

**Comment by jleandroperez** [comment]
@diegoreymendez simply log into a `WordPress.com` account that has 2FA enabled. The backend should send a Push Notification which should be translated into an alertView in the device.

If you approve the login request, WPiOS is expected to hit the backend and authorize the web login-attempt. Please, note that you may need to run the Push Notifications Development Server (if anything, please, poke me over Slack!)

**Comment by diegoreymendez** [comment]
@jleandroperez - I'd like another approval before merging this in (at least two).  Would you mind taking a look at my reply to @koke's comments and at the PR in general and let me know if it looks good?

**Comment by SergioEstevao** [comment]
@diegoreymendez  I noticed that this test is failing:

```
func testAuthorizeLoginUsesTheCorrectPath() {
        pushAuthenticationServiceRemote?.authorizeLogin(token, success: nil, failure: nil)
        
        XCTAssertTrue(mockRemoteApi!.postMethodCalled, "Method was not called")
        XCTAssertEqual(mockRemoteApi!.URLStringPassedIn!, "me/two-step/push-authentication", "Incorrect URL passed in")
    }
``` on PushAuthenticationServiceRemoteTests because the endpoint version is part of the URL this assert  ```XCTAssertEqual(mockRemoteApi!.URLStringPassedIn!, "me/two-step/push-authentication", "Incorrect URL passed in")``` fails.

**Comment by diegoreymendez** [comment]
@SergioEstevao - Good catch.  The unit test is now fixed and ready for another look.

**Comment by diegoreymendez** [comment]
@koke - Addressed the `.location == 0` request.

The initial reason I didn't go that route is that this change breaks requests that provide a full paths, although I believe there are no longer such requests and they don't make sense anyway.

**Comment by koke** [comment]
I like the changes and everything seems to work. Good work on this one :clap: :shipit:

**Comment by diegoreymendez** [comment]
Thanks @koke!  I'll be reviewing a comment @jleandroperez sent me before merging.

**Comment by diegoreymendez** [comment]
So @jleandroperez found out some services were staying out of the change.  I'll be fixing those now.

Just as a reference: they're the services inside `WordPressComApi.m`.

Thanks Jorge!

**Comment by diegoreymendez** [comment]
@koke - Had to roll back the `.location == 0` change proposed above, since some services in the App are using full paths.  For example [this](https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/4120-dynamic-endpoint-version/WordPress/Classes/Services/ReaderPostService.m#L62).

**Comment by diegoreymendez** [comment]
@jleandroperez - This is ready for another look.

**Comment by jleandroperez** [comment]
@diegoreymendez awesome job!!. Just tested the updated Swift services (along with... as much as i could think of).

:shipit: on my side!

**Comment by diegoreymendez** [comment]
Ty @jleandroperez !
